File: 2015_asym_sts/megha/ParseFile.m
function [Nsubj, Ncol, SubjID, Data, PhenoNames] = ParseFile(filename, delimiter, header)
% This function parses a plain text file with specified delimiter with no missing values.
%
% Input arguments - 
% filename: directory and name of the text file
% Col 1: family ID; Col 2: subject ID; From Col 3: numerical values
% delimiter: delimiter used in the text file
% header: 1 - PhenoFile contains a headerline
% 0 - PhenoFile does not contain a headerline; default header = 0
%
% Output arguments -
% Nsubj: total number of subjects
% Ncol: number of quantitative variables
% SubjID: a list of subject IDs
% Data: an Nsubj x Ncol matrix for quantitative data
% PhenoNames: names of the phenotypes in the headerline

% check inputs
if nargin < 3
    header = 0;
end
 
delimiter = sprintf(delimiter);   % delimiter

fid = fopen(filename);   % open the file

SubjID = textscan(fid, '%*s %s %*[^\n]', 'Delimiter', delimiter, 'HeaderLines', header);   % read the second column (subject ID)
SubjID = SubjID{1};   % convert a 1x1 cell structure into a Nsubj x 1 cell structure
Nsubj = length(SubjID);   % calculate number of subjects

frewind(fid)   % set the file position indicator to the beginning of the file
line = fgetl(fid);   % read the first line of the data
Ncol = numel(strfind(line, delimiter)) - 1;   % calculate the number of columns for quantitative data

frewind(fid)   % set the file position indicator to the beginning of the file
Data = cell2mat(textscan(fid, ['%*s %*s', repmat('%f',1,Ncol)], 'Delimiter', delimiter, 'HeaderLines', header));   % read quantitative data

frewind(fid)   % set the file position indicator to the beginning of the file
if header == 1    
    PhenoNames = textscan(fid, ['%*s %*s', repmat('%s',1,Ncol)], 1, 'Delimiter', delimiter);   % get the names of the phenotypes
else
    PhenoNames = 'NA';   % no headerline
end

fclose(fid);   % close the file
%

File: 2016_sulcal_depth/megha/R/ParseFile.R
ParseFile <- function(filename, delimiter="\t", header=FALSE){
# This function parses a plain text file with specified delimiter with no missing values.
	
	ID <- read.table(pipe(paste("cut -f2", filename)), header=header, sep=delimiter, comment.char="")   # read the second column (subject ID)
	SubjID = ID[[1]]   # extract subject IDs
	Nsubj <- length(SubjID)   # calculate the number of subjects
	
	line <- read.table(pipe(paste("cut -f3-", filename)), header=header, sep=delimiter, nrow=1, comment.char="")   # read the first line of the file from the third column
	Ncol = ncol(line)   # calculate the number of columns for quantitative data
	
	con <- pipe(paste("cut -f3-", filename))   # read the file from the third column
	Data <- matrix(scan(con, what=numeric(), skip=as.integer(header), sep=delimiter, nlines=Nsubj, n=Nsubj*Ncol, comment.char=""), Nsubj, Ncol, byrow=TRUE)   # extract quantitative data
	close(con)   # close the file
	
	if (header){
		PhenoNames <- as.character(t(read.table(pipe(paste("cut -f3-", filename)), header=FALSE, sep=delimiter, nrow=1, comment.char="")))   # get the names of the phenotypes
	} else{
		PhenoNames <- "NA"   # no headerline
	}
	
	return(list(Nsubj=Nsubj, Ncol=Ncol, SubjID=SubjID, Data=Data, PhenoNames=PhenoNames))
}